From: Baolin Wang <baolin.wang@linux.alibaba.com>
Subject: mm-page_alloc-fix-freelist-movement-during-block-conversion-fix
Date: Fri, 5 Apr 2024 20:11:47 +0800

fix allocation failures with CONFIG_CMA

The original code logic was that if the 'migratetype' type allocation is
failed, it would first try CMA page allocation and then attempt to
fallback to other migratetype allocations.  Now it has been changed so
that if CMA allocation fails, it will directly return.  This change has
caused a regression when I running the thpcompact benchmark, resulting in
a significant reduction in the percentage of THPs like below:

thpcompact Percentage Faults Huge
                          K6.9-rc2                K6.9-rc2 + this patch
Percentage huge-1        78.18 (   0.00%)       42.49 ( -45.65%)
Percentage huge-3        86.70 (   0.00%)       35.13 ( -59.49%)
Percentage huge-5        90.26 (   0.00%)       52.35 ( -42.00%)
Percentage huge-7        92.34 (   0.00%)       31.84 ( -65.52%)
Percentage huge-12       91.18 (   0.00%)       45.85 ( -49.72%)
Percentage huge-18       89.00 (   0.00%)       29.18 ( -67.22%)
Percentage huge-24       90.52 (   0.00%)       46.68 ( -48.43%)
Percentage huge-30       94.44 (   0.00%)       38.35 ( -59.39%)
Percentage huge-32       93.09 (   0.00%)       39.37 ( -57.70%)

After making the following modifications, the regression is gone.

Link: https://lkml.kernel.org/r/a97697e0-45b0-4f71-b087-fdc7a1d43c0e@linux.alibaba.com
Signed-off-by: Baolin Wang <baolin.wang@linux.alibaba.com>
Cc: David Hildenbrand <david@redhat.com>
Cc: "Huang, Ying" <ying.huang@intel.com>
Cc: Johannes Weiner <hannes@cmpxchg.org>
Cc: Mel Gorman <mgorman@techsingularity.net>
Cc: Vlastimil Babka <vbabka@suse.cz>
Cc: Zi Yan <ziy@nvidia.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/page_alloc.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/mm/page_alloc.c~mm-page_alloc-fix-freelist-movement-during-block-conversion-fix
+++ a/mm/page_alloc.c
@@ -2139,7 +2139,8 @@ __rmqueue(struct zone *zone, unsigned in
 	if (unlikely(!page)) {
 		if (alloc_flags & ALLOC_CMA)
 			page = __rmqueue_cma_fallback(zone, order);
-		else
+
+		if (!page)
 			page = __rmqueue_fallback(zone, order, migratetype,
 						  alloc_flags);
 	}
_
